agent/proxycfg/state.go

```diff
@@ -0,0 +1,357 @@
+package proxycfg
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/mitchellh/copystructure"
+)
+
+const coallesceTimeout = 200 * time.Millisecond
+
+// State holds all the state needed to maintain the config for a registeres
+// connect-proxy service. When a proxy registration is changed, the entire state
+// is discarded and a new one created.
+type State struct {
+	ctx    context.Context
+	cancel func()
+
+	proxyID  string
+	address  string
+	port     int
+	proxyCfg structs.ConnectProxyConfig
+	token    string
+
+	ch chan cache.UpdateEvent
+
+	snapCh chan ConfigSnapshot
+	reqCh  chan chan *ConfigSnapshot
+
+	logger *log.Logger
+	source *structs.QuerySource
+}
+
+// ConfigSnapshot captures all the resulting config needed for a proxy instance.
+// It is meant to be point-in-time coherent and is used to deliver the current
+// config state to observers who need it to be pushed in (e.g. XDS server).
+type ConfigSnapshot struct {
+	ProxyID           string
+	Address           string
+	Port              int
+	Proxy             structs.ConnectProxyConfig
+	Roots             *structs.IndexedCARoots
+	Leaf              *structs.IssuedCert
+	UpstreamEndpoints map[string]structs.CheckServiceNodes
+
+	// Skip intentions for now as we don't push those down yet, just pre-warm them.
+}
+
+// Valid returns whether or not the snapshot has all required fields filled yet.
+func (s *ConfigSnapshot) Valid() bool {
+	return s.Roots != nil && s.Leaf != nil
+}
+
+// Clone makes a deep copy of the snapshot we can send to other goroutines
+// without worrying that they will racily read or mutate shared maps etc.
+func (s *ConfigSnapshot) Clone() (*ConfigSnapshot, error) {
+	snapCopy, err := copystructure.Copy(s)
+	if err != nil {
+		return nil, err
+	}
+	return snapCopy.(*ConfigSnapshot), nil
+}
+
+// NewState populates the state from a NodeService struct. It will start up
+// watchers on any required resources for the proxy which are not released until
+// Close is called. The NodeService is assumed not to change for the lifetime of
+// the State; when it is, this state should be closed and a new one started.
+// It's also assumed that it's safe to read from it without coordination in the
+// current goroutine. The pointer is NOT preserved and accessed again later from
+// another goroutine. This is simpler than diffing states and the performance
+// overhead is not a big concern for now.
+func NewState(logger *log.Logger, c *cache.Cache, source *structs.QuerySource,
+	ns *structs.NodeService, token string) (*State, error) {
```

Comment:
Haha, same pattern here. I really prefer to avoid this, the API gets nasty really fast. I'd move to exported fields or a config struct.

Reply:
I would vote for the configuration struct route. That is the approach I have started taking for constructors that require more than 3 or 4 arguments.